src/pygeon/discretizations/fem/hdiv.py

```diff
@@ -411,42 +409,55 @@ def assemble_mass_matrix(
     def eval_basis_at_node(
         self,
         sd: pg.Grid,
-        dof_loc: np.ndarray,
-        cell_nodes_loc: np.ndarray,
+        opposites: np.ndarray,
         faces_loc: np.ndarray,
+        return_node_ind: bool = False,
     ) -> np.ndarray:
         """
         Compute the local basis function for the BDM1 finite element space.
 
         Args:
             sd (pg.Grid): The grid object.
-            dof_loc (np.ndarray): The local degrees of freedom.
+            opposites (np.ndarray): The local degrees of freedom.
             cell_nodes_loc (np.ndarray): The local nodes of the cell.
             faces_loc (np.ndarray): The local faces.
 
         Returns:
             np.ndarray: The local mass matrix.
         """
-        face_nodes_loc = sd.face_nodes[:, faces_loc].astype(bool).toarray()
+        fn = sd.face_nodes
+        nodes = np.empty((sd.dim + 1, sd.dim), int)
+        for ind, face in enumerate(faces_loc):
+            nodes[ind] = fn.indices[fn.indptr[face] : fn.indptr[face + 1]]
+        nodes = nodes.ravel(order="F")
 
-        # get the opposite node id for each face
-        opposite_node = np.logical_xor(face_nodes_loc, cell_nodes_loc)
+        node_ind = np.repeat(np.arange(sd.dim + 1), sd.dim)
 
-        # Find the nodes of the cell and their coordinates
-        indices = np.unique(dof_loc, return_inverse=True)[1].reshape((sd.dim, -1))
+        if not np.all(nodes[:: sd.dim][node_ind] == nodes):
+            node_ind = np.unique(nodes, return_inverse=True)[1]
+
+        face_ind = np.tile(np.arange(sd.dim + 1), sd.dim)
+
+        # get the opposite node id for each face
+        # opposite_node = sd.compute_opposite_nodes()
+        opposite_nodes = opposites[face_ind]
 
         # Compute a matrix Psi such that Psi[i, j] = psi_i(x_j)
-        Psi = np.empty((sd.dim * (sd.dim + 1), sd.dim + 1), np.ndarray)
-        for face, nodes in enumerate(indices.T):
-            tangents = (
-                sd.nodes[:, face_nodes_loc[:, face]]
-                - sd.nodes[:, opposite_node[:, face]]
-            )
-            normal = sd.face_normals[:, faces_loc[face]]
-            for index, node in enumerate(nodes):
-                val_at_node = tangents[:, index] / np.dot(tangents[:, index], normal)
-                Psi[face + index * (sd.dim + 1), node] = val_at_node.reshape((1, -1))
-        return sps.bmat(Psi)
+        tangents = sd.nodes[:, nodes] - sd.nodes[:, opposite_nodes]
+        normals = sd.face_normals[:, faces_loc[face_ind]]
+        vals = tangents / np.sum(tangents * normals, axis=0)
+
+        # Create a (i, j, v) triplet
+        dof_id = np.tile(np.arange(sd.dim * (sd.dim + 1)), 3)
+        nod_id = 3 * np.tile(node_ind, (3, 1)) + np.arange(3)[:, None]
+
+        result = np.zeros((sd.dim * (sd.dim + 1), 3 * (sd.dim + 1)))
+        result[dof_id, nod_id.ravel()] = vals.ravel()
+
+        if return_node_ind:
```

Comment:
when this option has been used?

Reply:
in assemble_lumped_matrix. I have added a comment in the doc-string